zmon_worker_monitor/zmon_worker/tasks/main.py

```diff
@@ -1632,39 +1632,43 @@ def _evaluate_downtimes(self, alert_id, entity_id):
-                    #     'alertId': alert_id,
-                    #     'entity': entity_id,
-                    #     'startTime': d['start_time'],
-                    #     'endTime': d['end_time'],
-                    #     'userName': d['created_by'],
-                    #     'comment': d['comment'],
-                    # })
-
-                # If downtime is over, we can remove its definition from redis.
-                if func == 'srem':
-                    if len(downtimes) == 1:
-                        p.delete('zmon:downtimes:{}:{}'.format(alert_id, entity_id))
-                        if len(redis_entities) == 1:
-                            p.delete('zmon:downtimes:{}'.format(alert_id))
-                            p.srem('zmon:downtimes', alert_id)
-                        else:
-                            p.srem('zmon:downtimes:{}'.format(alert_id), entity_id)
+                try:
+                    if now > d['start_time'] and now < d['end_time']:
+                        d['id'] = uuid
+                        result.append(d)
+                        func = 'sadd'
+                    elif now >= d['end_time']:
+                        func = 'srem'
                     else:
-                        p.hdel('zmon:downtimes:{}:{}'.format(alert_id, entity_id), uuid)
-                    p.execute()
+                        continue
+
+                    # Check whether the downtime changed state: active -> inactive or inactive -> active.
+                    changed = getattr(self.con, func)(
+                        'zmon:active_downtimes', '{}:{}:{}'.format(alert_id, entity_id, uuid))
+                    if changed:
+                        pass
+                        # eventlog.log(EVENTS[('DOWNTIME_ENDED' if func == 'srem' else 'DOWNTIME_STARTED')].id, **{
+                        #     'alertId': alert_id,
+                        #     'entity': entity_id,
+                        #     'startTime': d['start_time'],
+                        #     'endTime': d['end_time'],
+                        #     'userName': d['created_by'],
+                        #     'comment': d['comment'],
+                        # })
+
+                    # If downtime is over, we can remove its definition from redis.
+                    if func == 'srem':
+                        if len(downtimes) == 1:
+                            p.delete('zmon:downtimes:{}:{}'.format(alert_id, entity_id))
+                            if len(redis_entities) == 1:
+                                p.delete('zmon:downtimes:{}'.format(alert_id))
+                                p.srem('zmon:downtimes', alert_id)
+                            else:
+                                p.srem('zmon:downtimes:{}'.format(alert_id), entity_id)
+                        else:
+                            p.hdel('zmon:downtimes:{}:{}'.format(alert_id, entity_id), uuid)
+                        p.execute()
+                except:
```

Comment:
For me this looks ok, if the cleanup fails, it will happen on the next successful redis call? if redis is unavailable stuff fails anyway.

Important is only that an active alert executes and reports properly with switching to active in all failure cases to not hide the alert.